runtime/gc_vlhgc/CopyForwardCompactGroup.hpp

```diff
@@ -126,12 +132,122 @@ class MM_CopyForwardCompactGroup
 	{
 		_TLHRemainderBase = NULL;
 		_TLHRemainderTop = NULL;
+		_gcIDForTLHRemainder = 0;
 	}
 
 	MMINLINE void setTLHRemainder(void *base, void *top)
 	{
 		_TLHRemainderBase = base;
 		_TLHRemainderTop = top;
+		_TLHRemainderCount += 1;
+	}
+
+	MMINLINE void moveTLHRemainder(MM_CopyForwardCompactGroup *src)
+	{
+		_TLHRemainderBase = src->_TLHRemainderBase;
+		src->_TLHRemainderBase = NULL;
+		_TLHRemainderTop = src->_TLHRemainderTop;
+		src->_TLHRemainderTop = NULL;
+		_gcIDForTLHRemainder = src->_gcIDForTLHRemainder;
+		src->_gcIDForTLHRemainder = 0;
+	}
+
+	MMINLINE uintptr_t getTLHRemainderSize()
+	{
+		return ((uintptr_t)_TLHRemainderTop - (uintptr_t)_TLHRemainderBase);
+	}
+
+	/**
+	 * return true if TLHRemainder has been aligned.
+	 */
+	MMINLINE bool alignTLHRemainder(uintptr_t tlhSurvivorDiscardThreshold)
+	{
+		bool aligned = false;
+		if (_TLHRemainderBase) {
+			void *newAlignedTLHRemainderBase = (void*) MM_Math::roundToCeilingCard((uintptr_t) _TLHRemainderBase);
+			void *newAlignedTLHRemainderTop =  (void*) MM_Math::roundToFloorCard((uintptr_t) _TLHRemainderTop);
+			if (((uintptr_t)newAlignedTLHRemainderTop - (uintptr_t)newAlignedTLHRemainderBase) >= tlhSurvivorDiscardThreshold) {
+				/* TLHRemainder need to be card aligned for being reuse cross PGCs */
+				_TLHRemainderBase = newAlignedTLHRemainderBase;
+				_TLHRemainderTop = newAlignedTLHRemainderTop;
+				aligned = true;
+			}
+			/* else aligned TLHRemainder is smaller than tlhSurvivorDiscardThreshold, need to be reset */
+		}
+		return aligned;
+	}
+
+	MMINLINE bool abandonTLHRemainder(MM_EnvironmentVLHGC* env, uintptr_t compactGroup, bool preserveRemainder)
+	{
+		bool abandoned = true;
+		if (NULL != _TLHRemainderBase) {
+			Assert_MM_true(NULL != _TLHRemainderTop);
+			/* make it a walkable hole */
+			env->_cycleState->_activeSubSpace->abandonHeapChunk(_TLHRemainderBase, _TLHRemainderTop);
+			MM_GCExtensions* extensions =MM_GCExtensions::getExtensions(env);
+			MM_HeapRegionDescriptorVLHGC *region = (MM_HeapRegionDescriptorVLHGC*)extensions->heapRegionManager->tableDescriptorForAddress(_TLHRemainderBase);
+
+			/* only preserve for non Nursery regions */
+			if (!preserveRemainder || (MM_CompactGroupManager::getRegionAgeFromGroup(env, compactGroup) < extensions->tarokNurseryMaxAge._valueSpecified) || !region->containsObjects()) {
```

Comment:
in theory, we could simplify and preserve it even by nursery ones, since they will be cleared at the start of next PGC due to being a part of mandatory Collection Set (and just leave such a comment here instead of having the extra code)? either way is ok

Reply:
keep nursery check for avoid extra compactgroup up for preserved remainders, but remove preserveRemainder, using age instead

Reply:
hm, compact group and TLHRemainder fields already exist even for nursery ages. not sure what you are saving?